cfg/builder/builder_finalize.cc

```diff
@@ -130,92 +131,102 @@ void CFGBuilder::sanityCheck(core::Context ctx, CFG &cfg) {
     }
 }
 
-core::LocalVariable maybeDealias(core::Context ctx, core::LocalVariable what,
-                                 UnorderedMap<core::LocalVariable, core::LocalVariable> &aliases) {
-    if (what.isSyntheticTemporary(ctx)) {
-        auto fnd = aliases.find(what);
+LocalRef maybeDealias(core::Context ctx, CFG &cfg, LocalRef what, UnorderedMap<LocalRef, LocalRef> &aliases) {
+    if (what.isSyntheticTemporary(ctx, cfg)) {
+        auto fnd = aliases.find(what.id());
         if (fnd != aliases.end()) {
             return fnd->second;
-        } else {
-            return what;
         }
-    } else {
-        return what;
     }
+    return what;
 }
 
 /**
  * Remove aliases from CFG. Why does this need a separate pass?
  * because `a.foo(a = "2", if (...) a = true; else a = null; end)`
  */
 void CFGBuilder::dealias(core::Context ctx, CFG &cfg) {
-    vector<UnorderedMap<core::LocalVariable, core::LocalVariable>> outAliases;
-
+    vector<UnorderedMap<LocalRef, LocalRef>> outAliases;
     outAliases.resize(cfg.maxBasicBlockId);
+
     for (auto it = cfg.forwardsTopoSort.rbegin(); it != cfg.forwardsTopoSort.rend(); ++it) {
         auto &bb = *it;
         if (bb == cfg.deadBlock()) {
             continue;
         }
-        UnorderedMap<core::LocalVariable, core::LocalVariable> &current = outAliases[bb->id];
+        auto &current = outAliases[bb->id];
+
         if (!bb->backEdges.empty()) {
+            // Take the intersection of all of the back edges' aliases.
```

Comment:
This comment helps explain the code changes below.

The previous code did the following:

* Block enters loop with an empty set of aliases.
* If block has at least one backedge:
  * Set the current block's aliases to the aliases of the first backedge.
* Take the intersection of current block's aliases and all of its backedges' aliases.

The new code does the following:

* Block enters loop with an empty set of aliases.
* If block has at least one backedge:
  * Set the current block's aliases to the aliases of the first backedge.
  * Take the intersection of current block's aliases and all of its backedges' aliases except for the first one.

The change:

* Removes the loop from the path of blocks w/o backedges (a common case)
* For blocks with backedges, avoids a no-op loop that intersects two identical sets.
  * This results in noticeable savings since many blocks only have a single backedge.